lib/puppet/resource_api/property.rb

```diff
@@ -69,10 +86,40 @@ def rs_value
   # method overloaded only for the :ensure property, add option to check if the
   # rs_value matches is. Only if the class is child of
   # Puppet::ResourceApi::Property.
-  def def_insync?
+  def def_ensure_insync?
     define_singleton_method(:insync?) { |is| rs_value.to_s == is.to_s }
   end
 
+  def def_custom_insync?
+    define_singleton_method(:insync?) do |is|
+      provider    = @referrable_type.my_provider
+      context     = @referrable_type.context
+      should_hash = @resource.rsapi_canonicalized_target_state
+      is_hash     = @resource.rsapi_current_state
+      title       = @resource.rsapi_title
+
+      raise(Puppet::DevError, 'No insync? method defined in the provider; an insync? method must be defined if the custom_insync feature is defined for the type') unless provider.respond_to?(:insync?)
+      provider_insync_result = provider.insync?(context, title, @attribute_name, is_hash, should_hash)
+      # if nil, super and rely on Puppet::Property.insync?
+      # if true/false, return that and use the default change messaging
+      # if string, set value of custom change reporting and treat as false
+      @change_to_s_value = provider_insync_result if provider_insync_result.is_a?(String)
+      return provider_insync_result == true unless provider_insync_result.nil?
+
+      super(is)
+    end
+
+    define_singleton_method(:change_to_s) do |current_value, newvalue|
+      # As defined in the custom insync? method, it is sometimes useful to overwrite the default change messaging;
+      # The enables a user to return a more useful change report than a strict "is to should" report.
+      # If @change_to_s_value is not set, Puppet writes a generic change notification, like:
+      #   Notice: /Stage[main]/Main/<type_name>[<name_hash>]/<property name>: <property name> changed <is value> to <should value>
+      # If #change_to_s_value is *nil* Puppet writes a weird empty message like:
+      #   Notice: /Stage[main]/Main/<type_name>[<name_hash>]/<property name>:
+      @change_to_s_value || super(current_value, newvalue)
```

Comment:
It seems klunky for the `insync?` method to pass back a "changed" message when [elsewhere in the resource-api docs we say](https://puppet.com/docs/puppet/7/about_the_resource_api.html#resource_api_runtime_environment-logging-reporting-utilities)

> The provider needs to signal changes, successes, and failures to the runtime environment. The context is the primary way to do this. It provides a structured logging interface for all provider actions. 

Couldn't the provider just override the change_to_s behavior, for properties that need it, such as a password:

```ruby
def set(context, changes)
  changes.each do |change|
    if change[:name] == 'password'
      context.attribute_changed(change[:title], change[:name], old, new, "Changed [redacted] to [redacted]")
    else
      context.attribute_changed(change[:title], change[:name], old, new) # use default message
     end
  end
  do_something unless noop?
end
```

Reply:
Your assessment is spot on. This is not great. Sadly, the type/provider is not being called in a way that this would actually be implementable.

Puppet first sets all property values that need it, based on `insync?` logic, including calling into `change_to_s` and emits log messages accordingly. Only later `set` (via `flush`) is called.

I mean, I could imagine decoupling the two implementations completely and calling `set` the first time `insync?` is called, store all change notifications and other logs and dolly them out as puppet catches up with reality, but that sounds like a recipe for even more magic and disaster (where log-messages, actions and failures are entirely disconnected from each other).

Reply:
> Puppet first sets all property values that need it, based on insync? logic, including calling into change_to_s and emits log messages accordingly

I don't think that's true. The puppet type/provider API calls `change_to_s` during the `sync` phase not `insync?` This can be seen in the backtrace:

```
--> #0  Puppet::Property.change_to_s(current_value#Symbol, newvalue#String) at /home/josh/work/puppet/lib/puppet/property.rb:204
    #1  Puppet::Transaction::ResourceHarness.sync(event#Puppet::Transaction::Event, param#Puppet::Type::Notify::Message, current_value#Symbol, audit_message#String) at /home/josh/work/puppet/lib/puppet/transaction/resource_harness.rb:247
```

Reply:
I believe you mean https://github.com/puppetlabs/puppet/blob/3aecb85fd55235d0d5f98170d3268a756f000a17/lib/puppet/transaction/resource_harness.rb#L247 , right? 
This is exactly when `param.sync` (L241) is called and is only called itself if `param.safe_insync?` (L125) returns false.

All of rsapi's action happens in `flush` (https://github.com/puppetlabs/puppet/blob/3aecb85fd55235d0d5f98170d3268a756f000a17/lib/puppet/transaction/resource_harness.rb#L25) after all params have been synced and change log messages already emitted (L21, L89) - and with no way to suppress the change message.

Also note that `flush` only is called if there is at least one change on the resource (L23) (and there is no `noop` in play).

Altogether I don't see a way to use the context as a conduit for this messaging without a significantly larger change to how puppet applies transactions.

Reply:
Ah I was confused when you said "based on insync? logic", as I took that literally to mean `provider.insync?` calls into `change_to_s`. But I see you're saying in general the per-property insync/sync logic calls into `change_to_s`, and that happens before flush.